crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+        let fe = unsafe { self.get_entry(fd)? };
+        let po_path = fe.preopen_path.as_ref().ok_or(Errno::Notsup)?;
+        if fe.file_type != types::Filetype::Directory {
+            return Err(Errno::Notdir);
+        }
+
+        let path = path::from_host(po_path.as_os_str())?;
+        let prestat = types::PrestatDir {
+            pr_name_len: path.len().try_into()?,
+        };
+        Ok(types::Prestat::Dir(prestat))
+    }
+
+    fn fd_prestat_dir_name(
+        &self,
+        fd: types::Fd,
+        path: GuestPtr<u8>,
+        path_len: types::Size,
+    ) -> Result<()> {
+        trace!(
+            "fd_prestat_dir_name(fd={:?}, path={:?}, path_len={})",
+            fd,
+            path,
+            path_len
+        );
+
+        // TODO: should we validate any rights here?
+        let fe = unsafe { self.get_entry(fd)? };
+        let po_path = fe.preopen_path.as_ref().ok_or(Errno::Notsup)?;
+        if fe.file_type != types::Filetype::Directory {
+            return Err(Errno::Notdir);
+        }
+
+        let host_path = path::from_host(po_path.as_os_str())?;
+        let host_path_len = host_path.len().try_into()?;
+
+        if host_path_len > path_len {
+            return Err(Errno::Nametoolong);
+        }
+
+        trace!("     | (path_ptr,path_len)='{}'", host_path);
+
+        let mut bc = GuestBorrows::new();
+        let path_arr = path.as_array(host_path_len);
+        let slice = path_arr.as_raw(&mut bc)?;
+        unsafe { &mut *slice }.copy_from_slice(host_path.as_bytes());
+
+        Ok(())
+    }
+
+    fn fd_pwrite(
+        &self,
+        fd: types::Fd,
+        ciovs: &types::CiovecArray<'_>,
+        offset: types::Filesize,
+    ) -> Result<types::Size> {
+        trace!(
+            "fd_pwrite(fd={:?}, ciovs={:?}, offset={})",
+            fd,
+            ciovs,
+            offset,
+        );
+
+        let mut slices = Vec::new();
+        let mut total_len: types::Size = 0;
+        let mut bc = GuestBorrows::new();
+        bc.borrow_slice(ciovs)?;
+        for ciov_ptr in ciovs.iter() {
+            let ciov_ptr = ciov_ptr?;
+            let ciov: types::Ciovec = ciov_ptr.read()?;
+            let buf = ciov.buf;
+            let buf_len = ciov.buf_len;
+            total_len = total_len.checked_add(buf_len).ok_or(Errno::Inval)?;
+            let as_arr = buf.as_array(buf_len);
+            let as_raw = as_arr.as_raw(&mut bc)?;
+            slices.push(unsafe { &*as_raw });
+        }
+
+        let entry = unsafe { self.get_entry(fd)? };
+        let file = entry
+            .as_descriptor(
+                types::Rights::FD_WRITE | types::Rights::FD_SEEK,
+                types::Rights::empty(),
+            )?
+            .as_file()?;
+
+        if offset > i64::max_value() as u64 {
+            return Err(Errno::Io);
+        }
+
+        let mut buf = Vec::with_capacity(total_len as usize);
+        for slice in &slices {
+            buf.extend_from_slice(slice);
```

Comment:
I'd have the same question as before with `pread` here about passing raw wasm buffers to the underlying implementation.

Reply:
Excellent observation, and as I explained above, I've only made enough to make `wiggle` work with `wasi-common`. I'm happy to investigate this further though!

Reply:
I've simplified this now, so have a look and lemme know what you reckon!